api/oci/ociutils/ref.go

```diff
@@ -0,0 +1,83 @@
+package ociutils
+
+import (
+	"strings"
+
+	"github.com/mandelsoft/goutils/generics"
+	"github.com/opencontainers/go-digest"
+)
+
+// ParseVersion parses the version part of an OCI reference consisting
+// of an optional tag and/or digest.
+func ParseVersion(vers string) (ArtVersion, error) {
+	if strings.HasPrefix(vers, "@") {
+		dig, err := digest.Parse(vers[1:])
+		if err != nil {
+			return ArtVersion{}, err
+		}
+		return ArtVersion{
+			Tag:    nil,
+			Digest: &dig,
+		}, nil
+	}
+
+	i := strings.Index(vers, "@")
+	if i > 0 {
+		dig, err := digest.Parse(vers[i+1:])
+		if err != nil {
+			return ArtVersion{}, err
+		}
+		return ArtVersion{
+			Tag:    generics.Pointer(vers[:i]),
+			Digest: &dig,
+		}, nil
+	}
+	return ArtVersion{
+		Tag: &vers,
+	}, nil
+}
+
+// ArtVersion is the version part of an OCI reference consisting of an
+// optional tag and/or digest. Both parts may be nil, if a reference
+// does not include a version part.
+type ArtVersion struct {
+	// +optional
+	Tag *string `json:"tag,omitempty"`
+	// +optional
+	Digest *digest.Digest `json:"digest,omitempty"`
+}
+
+func (v *ArtVersion) VersionSpec() string {
+	vers := ""
+	if v != nil {
+		if v.Tag != nil {
+			vers = *v.Tag
+		}
+		if v.Digest != nil {
+			vers += "@" + string(*v.Digest)
+		}
+		if vers == "" {
+			return "latest"
+		}
+	}
+	return vers
+}
+
+func (v *ArtVersion) IsVersion() bool {
```

Comment:
I think a method on a type called version with `IsVersion` is unintuitive.

```suggestion
func (v *ArtVersion) IsLatest() bool {
      return v != nil && v.Tag == nil && v.Digest == nil
}
```
then wherever you want to use IsVersion just use the inverse of IsLatest

Reply:
I just noticed that all methods except for VersionSpec cannot deal with being nil. Is that intended?

Reply:
IsVersion is used as embedded type in ociArtSpec to be compatible here.

It returns whether this object (reference) (or nil)  specifies a dedicated version.

added check for nil

Reply:
Understood regarding the reuse. can we write a short comment on this? Then im totally fine with keeping it as is

Reply:
done